### PR TITLE
[Retail][Resto Shaman] Season 4 Tier Set

### PR DESCRIPTION
**Pull request by lspinheiro**
### Description

This Pull request adds the Season 4 Tier set for Resto Shamans. It also updates `calculateEffectiveHealingFromCritIncrease` to possibly account for increated healing critical contribution based on the 4pc effect in the Tier.

### Testing

- Test report URL: `/report/mCchtJDQkbHnTyN7`
image

### CODE REVIEW

File: src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        size="flexible"
+        tooltip={`
+          During totem uptime:
+          Total Healing: ${formatNumber(this.totalHealingDuringTotemUptime)}
+          Increased Healing from 15% crit chance bonus: ${formatNumber(this.increasedCritHealing)} (${formatPercentage(increasedCritPercentage)}% of total healing)
+          Estimated additional crits: ${formatNumber(this.totalCritsFromBonus)}
+          ${this.has4pc ? `Increased Healing from 4pc bonus: ${formatNumber(this.increased4pcHealing)} (${formatPercentage(increased4pcPercentage)}% of total healing)` : ''}
+        `}
+      >
+        <div className="pad boring-text">
```
Comment (emallson) [inline src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: this should use the `BoringItemSetValue` component

File: src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ *
+ * 2pc: While [Cloudburst Totem / Healing Stream Totem] is active, your chance to critically strike is increased by 15%.
+ *
+ * 4pc : Your critical heals have 220% effectiveness instead of the usual 200%.
+ */
+enum ActiveTotem {
+  Cloudburst = 'cloudburst',
+  HealingStream = 'healing_stream',
+  None = 'none',
+}
```
Comment (emallson) [inline src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: to confirm: Cloudburst Totem replaces Healing Stream Totem, right? (i think it does, judging by the spell description, but CBT doesn't actually *say* it replaces HST)

File: src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        const actualCritChance = normalCritChance + CRIT_RATE_TOTEM_HEALING_INCREASE;
+        const chanceOfThisCritDueToBonus = (actualCritChance - normalCritChance) / actualCritChance;
+
+        this.increasedCritHealing += totalHealing * chanceOfThisCritDueToBonus;
+        this.totalCritsFromBonus += chanceOfThisCritDueToBonus;
+
+        if (this.has4pc) {
+          const critBonus =
+            this.critEffectBonus.getBonus(event) - CritEffectBonus.BASE_CRIT_EFFECT_MOD;
+          this.increased4pcHealing += totalHealing * (critBonus / 2);
```
Comment (emallson) [inline src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: we have an existing helper for the crit % increase ( `calculateEffectiveHealingFromCritIncrease`) that handles some edge cases around absorbed healing and overhealing that you probably ought to use

**Review by emallson — CHANGES_REQUESTED** [review]
hey! this looks like a solid first PR. i've got some small comments but overall seems good.

you'll also want to add yourself to the `CONTRIBUTORS` file and add a `CHANGELOG` entry for the resto spec

**Comment by lspinheiro** [comment]
> hey! this looks like a solid first PR. i've got some small comments but overall seems good.
> 
> you'll also want to add yourself to the `CONTRIBUTORS` file and add a `CHANGELOG` entry for the resto spec

Hey! Thanks for the great feedback @emallson . I opened as a draft to get some initial feedback from Arlie (just noticed I dont know her github user). I will make the updates, do some additional testing and try to make it ready for review later today.

File: src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx (hunk 2 above)
Comment (lspinheiro) [reply src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: It does. I don't generally use CBT and last time I used it didn't replace HST, but I just tested today and it does. I may be able to simplify this.

File: src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      Events.applybuff.by(SELECTED_PLAYER).spell(TALENTS_SHAMAN.CLOUDBURST_TOTEM_TALENT),
+      this.onCloudburstApply,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(TALENTS_SHAMAN.CLOUDBURST_TOTEM_TALENT),
+      this.onCloudburstRemove,
+    );
+
+    this.addEventListener(
+      Events.summon.by(SELECTED_PLAYER).spell(TALENTS_SHAMAN.HEALING_STREAM_TOTEM_SHARED_TALENT),
```
Comment (lspinheiro) [inline src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: @emallson , I'm having a hard time tracking when a totem is active. I have tried both `Events.applybuff` and `Events.summon` and neither work. Not sure if the issue is with detecting the talent or something else. I can see that in the warcraft logs the events I want is when the totem is summoned and when it dies. Any tips on how do I detect those events here?

image



image

Comment (emallson) [reply src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: can you link the log? odds are that the spell id is not the same as the spell listed in the talent. often for combining / charge granting talents the id is different

Comment (lspinheiro) [reply src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: Here is the log for the event. The `guid` is different both the spell and talent entries of "Healing Stream Totem". I could only find a reference to this guid in the classic components. Right now I'm checking against the hability name and it seems to work.

```json
ability: {name: 'Healing Stream Totem', guid: 5394, type: 8, abilityIcon: 'inv_spear_04.jpg'},
fight: 1,
sourceID: 1,
sourceIsFriendly: true,
targetID: 41,
targetInstance: 1,
targetIsFriendly: true,
timestamp: 647009,
type: "summon"
}

Comment (emallson) [reply src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: this is a case where you need to add the spell definition to `common/SPELLS/shaman` with the correct id/name/icon

Comment (emallson) [reply src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: often, spells have multiple effects and we put any that aren't in the auto-generated data under `SPELLS`

Comment (lspinheiro) [reply src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]: Got it. I will need to get some logs with cloudburst as well to check the correct spell id. I will update it soon

File: src/common/TALENTS/shaman.ts [hunk 5]
```diff
@@ -678,7 +678,7 @@ const talents = {
     manaCost: 10800,
   },
   HEALING_STREAM_TOTEM_RESTORATION_TALENT: {
-    id: 392916,
+    id: 5394,
```
Comment (emallson) [inline src/common/TALENTS/shaman.ts]: this is an auto-generated file and this modification won't be preserved. you need to add an entry to `common/SPELLS/shaman.ts` that looks like:

```ts
HEALING_STREAM_TOTEM: {
  id: 5394,
  name: 'Healing Stream Totem',
  icon: 'inv_spear_04',
},
```

File: src/analysis/retail/shaman/restoration/modules/Abilities.ts [hunk 6]
```diff
@@ -41,7 +41,10 @@ class Abilities extends CoreAbilities {
         },
       },
       {
-        spell: TALENTS.HEALING_STREAM_TOTEM_SHARED_TALENT.id,
+        spell: [
+          TALENTS.HEALING_STREAM_TOTEM_SHARED_TALENT.id,
+          TALENTS.HEALING_STREAM_TOTEM_RESTORATION_TALENT.id,
+        ],
```
Comment (emallson) [inline src/analysis/retail/shaman/restoration/modules/Abilities.ts]: re: my other comment, this needs to be `SPELLS.HEALING_STREAM_TOTEM.id`,

File: src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx [hunk 7]
```diff
@@ -56,59 +49,36 @@ class Season4Tier extends Analyzer {
...
+    return `${targetID}-${targetInstance}`;
   }
 
-  private onCloudburstRemove() {
-    if (this.activeTotem === ActiveTotem.Cloudburst) {
-      console.log('Cloudburst Totem removed');
-      this.activeTotem = ActiveTotem.None;
+  onSummon(event: SummonEvent) {
+    if (
+      event.ability.guid === TALENTS_SHAMAN.HEALING_STREAM_TOTEM_RESTORATION_TALENT.id ||
```
Comment (emallson) [inline src/analysis/retail/shaman/restoration/modules/dragonflight/Season4Tier.tsx]:
```suggestion
      event.ability.guid === SPELLS.HEALING_STREAM_TOTEM.id ||
```

**Review by emallson — CHANGES_REQUESTED** [review]
the `TALENTS` files are auto-generated and your changes will be overwritten the next time we import changes, which means that the module will break

**Comment by emallson** [comment]
once the spell entry is addressed this is good to merge, tho

**Comment by lspinheiro** [comment]
> once the spell entry is addressed this is good to merge, tho

Just fixed it.

**Comment by emallson** [comment]
perfect!

**Pull request merged**